File: TutorialSamples/ControlSamplesWpf/ZoomPanControlSampleVB/ZoomPanScaleConverter.vb
Imports System
Imports System.Globalization

Imports System.Windows.Markup

''' <summary>
''' Used in MainWindow.xaml to converts a scale value to a percentage.
''' It is used to display the 50%, 100%, etc that appears underneath the zoom and pan control.
''' </summary>
<MarkupExtensionReturnType(GetType(Double))>
Public NotInheritable Class ZoomPanScaleConverter
    Inherits MarkupExtension
    Implements IValueConverter

    Public Overrides Function ProvideValue(ByVal serviceProvider As IServiceProvider) As Object
        Return Me
    End Function

    ''' <summary>
    ''' Convert a fraction to a percentage.
    ''' </summary>
    ''' <param name="value"></param>
    ''' <param name="targetType"></param>
    ''' <param name="parameter"></param>
    ''' <param name="culture"></param>
    ''' <returns></returns>
    Public Function Convert(ByVal value As Object, ByVal targetType As Type,
                            ByVal parameter As Object, ByVal culture As CultureInfo) _
                            As Object Implements IValueConverter.Convert
        Return CDbl(CInt((CDbl(value) * 100.0)))
    End Function

    ''' <summary>
    ''' Convert a percentage back to a fraction.
    ''' </summary>
    ''' <param name="value"></param>
    ''' <param name="targetType"></param>
    ''' <param name="parameter"></param>
    ''' <param name="culture"></param>
    ''' <returns></returns>
    Public Function ConvertBack(ByVal value As Object, ByVal targetType As Type,
                                ByVal parameter As Object, ByVal culture As CultureInfo) _
                                As Object Implements IValueConverter.ConvertBack
        Return CDbl(value) / 100.0
    End Function
End Class

File: Samples/WpfTestSvgControl/ZoomPanScaleConverter.cs
using System;
using System.Globalization;
using System.Windows.Data;
using System.Windows.Markup;

namespace WpfTestSvgControl
{
    /// <summary>
    /// Used in MainWindow.xaml to converts a scale value to a percentage.
    /// It is used to display the 50%, 100%, etc that appears underneath the zoom and pan control.
    /// </summary>
    [MarkupExtensionReturnType(typeof(double))]
    public sealed class ZoomPanScaleConverter : MarkupExtension, IValueConverter
    {
        public override object ProvideValue(IServiceProvider serviceProvider)
        {
            return this;
        }

        /// <summary>
        /// Convert a fraction to a percentage.
        /// </summary>
        /// <returns></returns>
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            // Round to an integer value whilst converting.
            return (double)(int)((double)value * 100.0);
        }

        /// <summary>
        /// Convert a percentage back to a fraction.
        /// </summary>
        /// <returns></returns>
        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            return (double)value / 100.0;
        }
    }
}
